cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -459,21 +455,48 @@ func (c *Client) addMember(name string, peerAddrs string, isLearner bool) ([]Mem
 	return ret, nil
 }
 
-func memberPromote(ctx context.Context, cli *clientv3.Client, learnerID uint64) error {
+func (c *Client) MemberPromote(learnerID uint64) error {
 	klog.V(1).Infof("[etcd] Promoting a learner as a voting member: %016x", learnerID)
 	// TODO: warning logs from etcd client should be removed.
 	// The warning logs are printed by etcd client code for several reasons, including
 	// 1. can not promote yet(no synced)
 	// 2. context deadline exceeded
 	// 3. peer URLs already exists
 	// Once the client provides a way to check if the etcd learner is ready to promote, the retry logic can be revisited.
-	_, err := cli.MemberPromote(ctx, learnerID)
-	if err == nil {
-		klog.V(1).Infof("[etcd] The learner was promoted as a voting member: %016x", learnerID)
-		return nil
+	var (
+		lastError error
+	)
+	err := wait.ExponentialBackoff(etcdBackoff, func() (bool, error) {
+		cli, err := clientv3.New(clientv3.Config{
+			Endpoints:   c.Endpoints,
+			DialTimeout: etcdTimeout,
+			DialOptions: []grpc.DialOption{
+				grpc.WithBlock(), // block until the underlying connection is up
+			},
+			TLS: c.TLS,
+		})
```

Comment:
Can we reuse the etcd client in https://github.com/kubernetes/kubernetes/pull/115038/files#diff-1671fef50296cdc263fe784ea570d008c1e123d15c2efb9aea27554a82080aaaL374?

It needs too many changes. Not sure.

Reply:
Sure, moved the client out of the backoff 👍🏻